marconi/src/Marconi/Index/EpochStakepoolSize.hs

```diff
@@ -0,0 +1,190 @@
+                    | succ e0 == e1 -> Just t
+                    | e0 == e1 -> Nothing
+                    | otherwise -> error $ "This should never happen: consequent epochs wider apart than by one: " <> show (e0, e1)
+                  _ -> error "This should never happen"
+              )
+
+-- | From LedgerState get epoch stakepool size: a mapping of pool ID
+-- to amount staked. We do this by getting the _pstatkeSet stake
+-- snapshot and then use _delegations and _stake to resolve it into
+-- the desired mapping.
+getStakeMap :: C.LedgerState -> Maybe (M.Map C.PoolId C.Lovelace)
+getStakeMap ledgerState' = case ledgerState' of
+  C.LedgerStateByron _st                  -> Nothing
+  C.LedgerStateShelley st                 -> fromState st
+  C.LedgerStateAllegra st                 -> fromState st
+  C.LedgerStateMary st                    -> fromState st
+  C.LedgerStateAlonzo st                  -> fromState st
+  -- TODO Pattern LedgerStateBabbage missing in cardano-node
+  -- https://github.com/input-output-hk/cardano-node/blob/release/1.35/cardano-api/src/Cardano/Api/LedgerState.hs#L252-L281,
+  -- swap this to a pattern when it appears.
+  C.LedgerState (O.LedgerStateBabbage st) -> fromState st
+  where
+    fromState
+      :: forall proto era c
+       . (c ~ LE.Crypto era, c ~ O.StandardCrypto)
+      => O.LedgerState (O.ShelleyBlock proto era)
+      -> Maybe (M.Map C.PoolId C.Lovelace)
+    fromState st = Just res
+      where
+        nes = O.shelleyLedgerState st :: SL.NewEpochState era
+
+        stakeSnapshot = Shelley._pstakeSet . Shelley.esSnapshots . Shelley.nesEs $ nes :: Shelley.SnapShot c
+        stakes = Shelley.unStake $ Shelley._stake stakeSnapshot :: VMap.VMap VMap.VB VMap.VP (LC.Credential 'LK.Staking c) (L.CompactForm L.Coin)
+
+        delegations :: VMap.VMap VMap.VB VMap.VB (LC.Credential 'LK.Staking c) (LK.KeyHash 'LK.StakePool c)
+        delegations = Shelley._delegations stakeSnapshot
+
+        res :: M.Map C.PoolId C.Lovelace
+        res = M.fromListWith (+) $ map swap $ P.catMaybes $ VMap.elems $
+          VMap.mapWithKey (\cred spkHash -> (\c -> (C.Lovelace $ coerce $ L.fromCompact c, f spkHash)) <$> VMap.lookup cred stakes) delegations
+
+        f :: (LK.KeyHash 'LK.StakePool c) -> C.PoolId
+        f xk = C.StakePoolKeyHash xk
+
+indexer
+  :: FilePath -> FilePath -> SQL.Connection
+  -> S.Stream (S.Of Event) IO r
+indexer conf socket dbCon =
+    CS.ledgerStates conf socket C.QuickValidation
+  & toEvents
+  & sqlite dbCon
+
+-- * Sqlite back-end
+
+-- | Consume a stream of events and write them to the database. Also
+-- passes events on after they are persisted -- useful to knwow when
+-- something has been persisted.
+sqlite :: SQL.Connection -> S.Stream (S.Of Event) IO r -> S.Stream (S.Of Event) IO r
+sqlite c source = do
+  lift $ SQL.execute_ c
+      "CREATE TABLE IF NOT EXISTS stakepool_delegation (poolId BLOB NOT NULL, lovelace INT NOT NULL, epochNo INT NOT NULL)"
```

Comment:
I think we may need a primary key on this table, or some indicator that make a ROW unique, besides the implicit SQLite Row-Id. 
We don't drop the table, we recreate if it doesn't exist, and start inserting it to it, so it is possible to insert the same events multiple times which I don't think is correct